providers/dns/cloudxns/cloudxns.go

```diff
@@ -0,0 +1,215 @@
+	}
+
+	return &DNSProvider{
+		apiKey:    apiKey,
+		secretKey: secretKey,
+	}, nil
+}
+
+// Present creates a TXT record to fulfil the dns-01 challenge.
+func (c *DNSProvider) Present(domain, token, keyAuth string) error {
+	fqdn, value, ttl := acme.DNS01Record(domain, keyAuth)
+	zoneID, err := c.getHostedZoneID(fqdn)
+	if err != nil {
+		return err
+	}
+
+	return c.addTxtRecord(zoneID, fqdn, value, ttl)
+}
+
+// CleanUp removes the TXT record matching the specified parameters.
+func (c *DNSProvider) CleanUp(domain, token, keyAuth string) error {
+	fqdn, _, _ := acme.DNS01Record(domain, keyAuth)
+	zoneID, err := c.getHostedZoneID(fqdn)
+	if err != nil {
+		return err
+	}
+
+	recordID, err := c.findTxtRecord(zoneID, fqdn)
+	if err != nil {
+		return err
+	}
+
+	return c.delTxtRecord(recordID, zoneID)
+}
+
+func (c *DNSProvider) getHostedZoneID(fqdn string) (string, error) {
+	type Data struct {
+		ID     string `json:"id"`
+		Domain string `json:"domain"`
+	}
+
+	authZone, err := acme.FindZoneByFqdn(fqdn, acme.RecursiveNameservers)
+	if err != nil {
+		return "", err
+	}
+
+	result, err := c.makeReauest("GET", "domain", nil)
+	if err != nil {
+		return "", err
+	}
+
+	var domains []Data
+	err = json.Unmarshal(result, &domains)
+	if err != nil {
+		return "", err
+	}
+
+	for _, data := range domains {
+		if data.Domain == authZone {
+			return data.ID, nil
+		}
+	}
+
+	return "", fmt.Errorf("Zone %s not found in cloudxns for domain %s", authZone, fqdn)
+}
+
+func (c *DNSProvider) findTxtRecord(zoneID, fqdn string) (string, error) {
+	result, err := c.makeReauest("GET", fmt.Sprintf("record/%s?host_id=0&offset=0&row_num=2000", zoneID), nil)
+	if err != nil {
+		return "", err
+	}
+
+	var records []cloudXNSRecord
+	err = json.Unmarshal(result, &records)
+	if err != nil {
+		return "", err
+	}
+
+	for _, record := range records {
+		if record.Host == acme.UnFqdn(fqdn) && record.Type == "TXT" {
+			return record.RecordID, nil
+		}
+	}
+
+	return "", fmt.Errorf("No existing record found for %s", fqdn)
+}
+
+func (c *DNSProvider) addTxtRecord(zoneID, fqdn, value string, ttl int) error {
+	id, err := strconv.Atoi(zoneID)
+	if err != nil {
+		return err
+	}
+
+	payload := cloudXNSRecord{
+		ID:     id,
+		Host:   acme.UnFqdn(fqdn),
+		Value:  value,
+		Type:   "TXT",
+		LineID: 1,
+		TTL:    ttl,
+	}
+
+	body, err := json.Marshal(payload)
+	if err != nil {
+		return err
+	}
+
+	_, err = c.makeReauest("POST", "record", body)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *DNSProvider) delTxtRecord(recordID, zoneID string) error {
+	_, err := c.makeReauest("DELETE", fmt.Sprintf("record/%s/%s", recordID, zoneID), nil)
+	return err
+}
+
+func (c *DNSProvider) hmac(url, date, body string) string {
+	sum := md5.Sum([]byte(c.apiKey + url + body + date + c.secretKey))
+	return hex.EncodeToString(sum[:])
+}
+
+func (c *DNSProvider) makeReauest(method, uri string, body []byte) (json.RawMessage, error) {

```

Comment:
This should be spelled `makeRequest` -- I understand how with a Chinese keyboard this could be difficult. :) But we would appreciate it!